---
title: "Sample Selection"
author: Szi Kay Leung
date: "`r Sys.Date()`"
output:
  rmdformats::material:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: true
    highlight: tango
    code_folding: hide
    fig_caption: true
    cards: true



---
```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE) #,fig.width=14, fig.height= 7)


# libraries
suppressMessages(library("janitor")) # set row to column names
suppressMessages(library("dplyr")) 
suppressMessages(library("plyr")) # rename factors in columns
suppressMessages(library("DT")) 
suppressMessages(library("ggplot2")) 
suppressMessages(library("wesanderson")) # colours 
suppressMessages(library("stringr")) # str_replace 
suppressMessages(library("ggrepel")) # geom_text_repel
suppressMessages(library("gridExtra")) #grid_arrange
suppressMessages(library("grid")) #grid_arrange

source("/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/Scripts/ad_bdr/Rmarkdown_Input.R")
```
# Aim

**To choose 30 BDR Samples for Targeted Iso-Seq by**    
- Braak-Stage, starting off first with 0 and 6  
- RINscore and concentration

Caveats: 
- Good mix of gender, ages and APOE genotype
- PMI 
- FACS selected 

## Input Files {.tabset .tabset-pills}

1. Phenotype Data from 96 samples - *Filename = 96_samples_all_v2.csv*  
2. RIN values from RNA extracted by G.Wheildon - *Filename = BDR_96_RNA_QC_2020-09-02 - 11-33-54-RNA_sampleTable.csv*
3. Samples selected for FACS - *Filename = 30samplesReFACS.csv*

```{r}
# Input files 
# 1. Phenotype_All information of core 96 samples (96_samples_all_v2.xlsx) 
# 2. RIN_All from RNA extracted by G.Wheildon of core 96 samples
input_dir <- "/gpfs/mrc0/projects/Research_Project-MRC148213/sl693/AD_BDR/Samples/" #first tab of 96_samples_all_v2.xlsx
Phenotype_All <-read.csv(paste0(input_dir,"96_samples_all_v2_singletab.csv"))[-1]
RIN_All <- read.table(paste0(input_dir,"BDR_96_RNA_QC_2020-09-02 - 11-33-54-RNA_sampleTable.csv"), as.is = T, sep = ",") %>% row_to_names(row_number = 1)
FACS_All <- read.table(paste0(input_dir,"FACS.csv"), as.is = T, sep = ",") %>% row_to_names(row_number = 1)  %>% mutate(FACS = "FACS_Yes")

# Data Wrangle QC
# Gender, ensure F = Female, M = Male for all values
Phenotype_All$Gender <- revalue(Phenotype_All$Gender, c("F" = "female", "M" = "male"))
# Braak Stage
Phenotype_All$Braak.Stage_numeric <- lapply(Phenotype_All$Braak.Stage, function(x) 
  if(x == "Braak tangle stage I "){"1"
} else if (x == "Braak tangle stage II "){"2"
} else if (x == "Braak tangle stage III "){"3"
} else if (x == "Braak tangle stage IV "){"4"
} else if (x == "Braak tangle stage V "){"5"
})


```

### Phenotype: 96 Samples
```{r}
datatable(Phenotype_All, caption = "Latest Phenotype data of 96 samples",options = list(pageLength = 5, scrollX = TRUE))
```

### Phenotype: RIN
```{r}
datatable(RIN_All, caption = "RIN Values from G.Wheildon",options = list(pageLength = 5, scrollX = TRUE))
```

### FACS: 32 samples
```{r}
datatable(FACS_All, caption = "FACS selected samples from J.Davies",options = list(pageLength = 5, scrollX = TRUE))
```

## QC 
Checking to make sure data in columns (with similar names) in phenotype data match

#### Age
Columns: Age vs Age.y: `r all(Phenotype_All$Age == Phenotype_All$Age.y)`  
All the rows except for one sample (BBN_id: BBN002.30920) match in ages:   
- Age for sample = `r setdiff(Phenotype_All$Age,Phenotype_All$Age.x)`  
- Age.x for sample = `r setdiff(Phenotype_All$Age.x,Phenotype_All$Age)`  
Columns: Age.x vs Age.y: `r all(Phenotype_All$Age.x == Phenotype_All$Age.y)`  

#### Gender 
Columns: Gender vs Sex.x: `r all(Phenotype_All$Gender == Phenotype_All$Sex.x)`

#### Braak Stage
Columns: BraakTangle_numeric vs Braak_tangle: `r all(Phenotype_All$BraakTangle_numeric == Phenotype_All$Braak_tangle)`  
All the rows except for one sample (BBN_id: BBN_9944) match in Braak Stages:  
- BraakTangle for sample = `r setdiff(Phenotype_All$Braak_tangle, Phenotype_All$BraakTangle_numeric)`  

# Merge samples by phenotype data and RIN

## Phenotype and RIN Values
Samples in Phenotype data, RIN data, and FACs data are merged by Sample ID (concatenation of BBN.ID and Brain.ID).  

```{r}
# Simpify and Merge Phenotype and RIN dataframes
Phenotype <- Phenotype_All %>% mutate(ID = paste0(BBN.ID,";",Brain.ID)) %>% select(ID, Institute, BraakTangle_numeric, Age.x, Sex.x, Plate, APOE_geno, PMI)
RIN <- RIN_All %>% mutate(ID = paste0(BBN.ID,";",Brain.ID)) %>% select(ID,RINe,conc)
FACS <- FACS_All %>% 
  # ensure brain_id and centre_id match that in phenotype by ensuring same non-character strings
  mutate(ID = paste0(BBN_ID,";",Centre_ID)) %>% mutate(ID = str_remove(ID, "-1")) %>%
  select(ID, FACS)

# merge by ID (concatenation of BBN.ID and Brain.ID)
Final <- merge(Phenotype, RIN, by = "ID", all = T)

# check samples from phenotype and RIN values match
Final$RNAPhenotype <- apply(Final, 1, function(x) 
  if(!is.na(x[["Institute"]]) & !is.na(x[["RINe"]])){"Match"
  } else if (!is.na(x[["Institute"]]) & is.na(x[["RINe"]])){"No_RIN_value"
  } else if (is.na(x[["Institute"]]) & !is.na(x[["RINe"]])){"No_Phenotype_Data"})

# list samples selected for FACS sorting
Final$FACS <- apply(Final, 1, function(x) ifelse(x[["ID"]] %in% FACS$ID, "FACS_Yes", "FACS_No"))

Final %>% group_by(RNAPhenotype, FACS) %>% tally()

datatable(Final, caption = "Merged dataframe of phenotype data and RIN scores by Sample ID",options = list(pageLength = 5, scrollX = TRUE))
```

- `r nrow(FACS_All)` samples in FACs data (selected by J.Davies) are in the core 96 of those with RNA extraction  = `r nrow(Final[Final$FACS == "FACS_Yes",]) == nrow(FACS_All) `

# All samples 

## {.tabset .tabset-pills}

### Age and Braak Stages 

Good range of age across Braak Stages. Of note:  
- Only 5 samples in Braak Stage 0, with clear distinction between Sex and Age  
- Unsurprisingly, many more Braak Stage 6 than Braak Stage 0 and also slightly older

```{r}
# age distribution
Final %>% 
  ggplot(., aes(x = as.factor(BraakTangle_numeric), y = as.numeric(Age.x))) + geom_point(aes(colour = Sex.x)) + 
  theme_classic() + 
  labs(x = "Braak Stages", y = "Age") + 
  scale_colour_discrete(name = "Sex", labels = c("Female", "Male"))

```

### RIN and Braak Stages 

Similar range of RINs across Braak Stages. 
- Use all 5 samples for Braak Stage 0 moving forward for Iso-Seq, and the top 11 samples from Braak Stage 1 (i.e. highest RIN)
- Use 16 samples from Braak Stage 6 (threshold at RIN 4)


```{r}
### RIN and Braakstages
Final %>% 
  ggplot(., aes(x = as.factor(BraakTangle_numeric), y = as.numeric(RINe))) + geom_point(aes(colour = Sex.x)) + 
  theme_classic() + 
  labs(x = "Braak Stages", y = "RIN") + 
  scale_colour_discrete(name = "Sex", labels = c("Female", "Male"))
```

# Extreme Samples: Braak Stage 0 and 6

## {.tabset .tabset-pills}

### RIN and Concentration
As expected, positive correlation of RIN to concentration for samples. Most samples with concentration > 100ng/uL (used in Mouse Targeted Iso-Seq prep)

```{r}
Final %>% 
  filter(BraakTangle_numeric %in% c("0","1","6")) %>%
  ggplot(., aes(x = as.numeric(conc), y = as.numeric(RINe), colour = as.factor(BraakTangle_numeric), shape = Institute)) + geom_point(size = 3) +
  theme_classic() + 
  labs(y = "RIN", x = "Concentration (ng/uL)") + 
  scale_colour_manual(name = "Braak Stages", labels = c("0", "6"), values = c(wes_palette("Royal1")[1], wes_palette("Royal1")[2], wes_palette("Royal1")[3]))
```

### APOE genotype 
Mix of APOE genotype across Braak Stage 0 and 6. Note an "NA" value for one of the samples

```{r}
# APOE genotype
Final %>% 
  filter(BraakTangle_numeric %in% c("0","6")) %>% 
  ggplot(., aes(x = APOE_geno, y = RINe,colour = as.factor(BraakTangle_numeric))) + geom_point(size = 3) + 
  theme_classic() + 
  labs(y = "RIN", x = "APOE Genotype") +
  scale_colour_manual(name = "Braak Stages", labels = c("0", "6"), values = c(wes_palette("Royal1")[1], wes_palette("Royal1")[2]))
```

### FACS 
Note: Some of the high RIN AD samples were not selected for FACS. However, not planning on integrating FACS with transcriptomics


```{r}

Final %>% 
  filter(BraakTangle_numeric %in% c("0","1","6")) %>% 
  ggplot(., aes(x = as.factor(BraakTangle_numeric), y = RINe, colour = FACS)) + geom_point(size = 3) +
  theme_classic() + 
  labs(y = "RIN", x = "Braak Stages") +
  scale_colour_manual(name = "FACS Selected", labels = c("No", "Yes"), values = c(wes_palette("Darjeeling1")[1], wes_palette("Darjeeling1")[2]))

```

### PMI
During upgrade, Lorna mentioned the importance of short PMI to avoid changes to splicing and gene expression ([Ferriera et al (2018)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5811508/#!po=8.33333)). 


As expected, higher RIN values correlated with lower PMI. However, not the way PMI is measured varies between institutes and conditions, thus not a primary factor to consider when determining samples to be selected
```{r}

Final %>% 
  filter(BraakTangle_numeric %in% c("0","1","6")) %>%
  ggplot(., aes(x = as.numeric(PMI), y = as.numeric(RINe), colour = as.factor(BraakTangle_numeric), shape = Institute)) + geom_point(size = 3) +
  theme_classic() + 
  labs(y = "RIN", x = "PMI (hours)") + 
  scale_colour_manual(name = "Braak Stages", labels = c("0","1","6"), values = c(wes_palette("Royal1")[1], wes_palette("Royal1")[4],wes_palette("Royal1")[2])) +
  geom_hline(yintercept=6, linetype="dashed", color = "green", size=0.2) 


```


### Age

The higher RIN AD samples (demaracated by the threshold RIN at 6) are older than the Control samples (shaded in blue). There are some matched AD samples (shaded in blue) but RIN is significantly lower

```{r}
Final %>% 
  filter(BraakTangle_numeric %in% c("0","1","6")) %>%
  ggplot(., aes(x = as.numeric(Age.x), y = as.numeric(RINe), colour = as.factor(BraakTangle_numeric))) + geom_point(size = 3) +
  theme_classic() + 
  labs(y = "RIN", x = "Age (years)") + 
  scale_colour_manual(name = "Braak Stages", labels = c("0","1","6"), values = c(wes_palette("Royal1")[1], wes_palette("Royal1")[2],wes_palette("Royal1")[4])) +
  geom_hline(yintercept=6, linetype="dashed", color = "green", size=0.2) +
  annotate("rect", xmin = 60, xmax = 78, ymin = 3.5, ymax = 6.0, alpha = .1,fill = "blue")

```

# Samples for Targeted Iso-Seq 
There are:  
- Braak Stage 0 = `r nrow(Final[Final$BraakTangle_numeric == "0",])` samples  
- Braak Stage 1 = `r nrow(Final[Final$BraakTangle_numeric == "1",])` samples 
- Braak Stage 6 = `r nrow(Final[Final$BraakTangle_numeric == "6",])` samples  

## {.tabset .tabset-pills}

### All BraakStage 0, 1, 6 Samples
```{r}
Final %>% 
  filter(BraakTangle_numeric %in% c("0","1","6")) %>% 
  datatable(., caption = "All Braak Stage 0, 1 and 6,",options = list(pageLength = 5, scrollX = TRUE)) %>% 
  formatStyle('BraakTangle_numeric',  backgroundColor = styleEqual(c(0, 1, 6), c(wes_palette("Royal1")[1], wes_palette("Royal1")[4],wes_palette("Royal1")[2])))
```


### Selected Samples 

Total Samples: 30 samples (Braak Stage 0 + 1 = 15 samples, Braak Stage 6 = 15 samples)
- All of Braak Stage 0 (n = 5 samples)
- 10 of Braak Stage 1 ranked by highest RIN Score 
- 15 of Brakk Stage 6 ranked by highest RIN score

1st run - All 5 Braak Stage 0 samples and 5 Braak Stage 6 samples (need to order 10th Pacbio barcoding sequence, only 9 currently in lab)  
- Braak Stage 6 samples ranked by highest RIN Score and thus the highest concentration first 


= Lowest RIN score for AD = 5.4; Lowest RIN score for Control = 3.9
- Only 5 controls to choose from, alos note lowest RIN score has no APOE genotype

```{r}

# filter by Braak Stage 1, 6 and order by RIN score, the top 10 and 15 respectively
BS_0 <- Final %>% filter(BraakTangle_numeric == "0") 
BS_1 <- Final %>% filter(BraakTangle_numeric == "1") %>% filter(rank(desc(as.numeric(RINe))) <= 10) 
BS_6 <- Final %>% filter(BraakTangle_numeric == "6") %>% filter(rank(desc(as.numeric(RINe))) <= 15)

# All_Samples
Selected <- bind_rows(BS_0, BS_1, BS_6) 

datatable(Selected, caption = "All Braak Stage 0 and 6, Partial Braak Stage 1",options = list(pageLength = 5, scrollX = TRUE)) %>% 
  formatStyle('BraakTangle_numeric',  backgroundColor = styleEqual(c(0,1,6), c(wes_palette("Royal1")[1], wes_palette("Royal1")[4], wes_palette("Royal1")[2])))

```

### RIN
```{r}
ggplot(Selected, aes(x = as.factor(BraakTangle_numeric), y = as.numeric(RINe))) + geom_point(aes(colour = Sex.x)) + 
  theme_classic() + 
  labs(x = "Braak Stages", y = "RIN") + 
  scale_colour_discrete(name = "Sex", labels = c("Female", "Male"))

```

### Age

Not completely matched between Control and AD but within the same range
```{r}
ggplot(Selected, aes(x = as.factor(BraakTangle_numeric), y = as.numeric(Age.x))) + geom_jitter(aes(colour = Sex.x),width = 0.15) + 
  theme_classic() + 
  labs(x = "Braak Stages", y = "Age (years)") + 
  scale_colour_discrete(name = "Sex", labels = c("Female", "Male"))

```

### PMI
One AD sample with very high PMI 
```{r}
ggplot(Selected, aes(x = as.factor(BraakTangle_numeric), y = as.numeric(PMI))) + geom_point(aes(colour = Sex.x)) + 
theme_classic() + 
labs(x = "Braak Stages", y = "Age") + 
scale_colour_discrete(name = "Sex", labels = c("Female", "Male"))
```

# Running in batches

## {.tabset .tabset-pills}

### Batch 1, 2, 3

```{r}

# 3 Batches in order of RIN scores (allow easier downstream Iso-Seq lab for amplification etc)
Batched <- bind_rows(arrange(Selected,desc(RINe)) %>% .[1:10,] %>% mutate(Batch = "1"),
          arrange(Selected,desc(RINe)) %>% .[11:20,] %>% mutate(Batch = "2"),
          arrange(Selected,desc(RINe)) %>% .[21:30,] %>% mutate(Batch = "3"))

# Manual Randomisation
Batched[Batched$ID == "BBN_18399;A053/11", "Batch"] <- "3"
Batched[Batched$ID == "BBN002.30920;A435/17", "Batch"] <- "2"

plot_batches <- function(batch_num){
  p <- Batched %>% filter(Batch == batch_num) %>% 
    ggplot(., aes(x = as.factor(BraakTangle_numeric), y = as.numeric(RINe), label = Plate, shape = Institute)) + 
    geom_jitter(aes(colour = Sex.x),width = 0.15, size = 2) + 
    theme_classic() + 
    labs(x = "Braak Stages", y = "RIN", subtitle = paste("Samples in Batch",batch_num)) + 
    scale_colour_discrete(name = "Sex", labels = c("Female", "Male")) + 
    #geom_text_repel(size = 3) + 
    ylim(3,8)
  
  return(p)
}

get_legend<-function(myggplot){
  tmp <- ggplot_gtable(ggplot_build(myggplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
}



# Significance across batches 
summary(aov(RINe ~ Batch, data = Batched))  # Statistically significant between Batches and RINe
summary(aov(RINe ~ Institute, data = Batched)) # Not signficant between RINe and Institute
summary(aov(Batch ~ Institute, data = Batched))# Nor signficant between Batch and Institute
write.csv(Batched, paste0(input_dir,"Selected_BDR.csv"))

legend <- get_legend(plot_batches("1"))
pdf(paste0(input_dir,"Selected_BDR_Batches.pdf"))
grid.arrange(plot_batches("1") + theme(legend.position="none"), 
             plot_batches("2") + theme(legend.position="none"), 
             plot_batches("3") + theme(legend.position="none"),
             legend, ncol=2)
dev.off()




